notebooks/working/ho_deconfounder_on_sim_data.py

```diff
@@ -0,0 +1,512 @@
+                          'ASC Shared Ride: 3+',
+                          'ASC Walk-Transit-Walk',
+                          'ASC Drive-Transit-Walk',
+                          'ASC Walk-Transit-Drive',
+                          'ASC Walk',
+                          'ASC Bike']
+
+mnl_specification["total_travel_time"] = [1, 2, 3, [4, 5, 6]]
+mnl_names["total_travel_time"] = ['Travel Time, units:min (Drive Alone)',
+                                  'Travel Time, units:min (SharedRide-2)',
+                                  'Travel Time, units:min (SharedRide-3+)',
+                                  'Travel Time, units:min (All Transit Modes)']
+
+mnl_specification["total_travel_cost"] = [1, 2, 3, [4, 5, 6]]
+mnl_names["total_travel_cost"] = ['Travel Cost, units:$ (Drive Alone)',
+                                  'Travel Cost, units:$ (SharedRide-2)',
+                                  'Travel Cost, units:$ (SharedRide-3+)',
+                                  'Travel Cost, units:$ (All Transit Modes)']
+
+# mnl_specification["cost_per_distance"] = [1, 2, 3]
+# mnl_names["cost_per_distance"] = ["Travel Cost per Distance, units:$/mi (Drive Alone)",
+#                                   "Travel Cost per Distance, units:$/mi (SharedRide-2)",
+#                                   "Travel Cost per Distance, units:$/mi (SharedRide-3+)"]
+
+mnl_specification["cars_per_licensed_drivers"] = [[1, 2, 3]]
+mnl_names["cars_per_licensed_drivers"] = ["Autos per licensed drivers (All Auto Modes)"]
+
+mnl_specification["total_travel_distance"] = [1, 2, 3, 7, 8]
+mnl_names["total_travel_distance"] = ['Travel Distance, units:mi (Drive Alone)',
+                                      'Travel Distance, units:mi (SharedRide-2)',
+                                      'Travel Distance, units:mi (SharedRide-3+)',
+                                      'Travel Distance, units:mi (Walk)',
+                                      'Travel Distance, units:mi (Bike)']
+
+# mnl_specification["cross_bay"] = [[2, 3], [4, 5, 6]]
+# mnl_names["cross_bay"] = ["Cross-Bay Tour (Shared Ride 2 & 3+)",
+#                           "Cross-Bay Tour (All Transit Modes)"]
+mnl_specification["cross_bay"] = [[2, 3]]
+mnl_names["cross_bay"] = ["Cross-Bay Tour (Shared Ride 2 & 3+)"]
+
+mnl_specification["household_size"] = [[2, 3]]
+mnl_names["household_size"] = ['Household Size (Shared Ride 2 & 3+)']
+
+mnl_specification["num_kids"] = [[2, 3]]
+mnl_names["num_kids"] = ["Number of Kids in Household (Shared Ride 2 & 3+)"]
+# -
+
+# # Deconfounder
+
+# ## Method 1
+#
+# In this method, we used all the variables in each mode's utility specification to recover its confounder, in line with Blei et al.'s approach in their paper.
+
+# +
+## Return the variables of each utility specification
+
+spec_dic = specifications(mnl_specification=mnl_specification, num_modes=8)
+spec_dic
+# -
+
+confounder_vectors = []
+holdout_dfs = []
+masks_df = []
+rows_df =[]
```

Comment:
What are each of these objects? What do they do / what is their purpose?
Can you add a comment above each variable answering these two questions?